### PR TITLE
Change storage api to not expose checksums directly

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
The storage API the platform is exposing does not support this functionality.  It does support asking if there is a key with a particular checksum, and that's all the functionality we need in roslyn itself.  So this switches to have the same form to make porting easier.

### CODE REVIEW

File: src/Workspaces/Core/Portable/Workspace/Solution/Checksum.cs [hunk 1]
```diff
@@ -136,6 +136,12 @@ public override unsafe string ToString()
         public static bool operator !=(Checksum left, Checksum right)
             => !(left == right);
 
+        public static bool operator ==(Checksum left, HashData right)
+            => left._checksum == right;
+
+        public static bool operator !=(Checksum left, HashData right)
+            => !(left == right);
```
Comment (sharwell) [inline src/Workspaces/Core/Portable/Workspace/Solution/Checksum.cs:143]: 😕 how/why is `HashData` not private?

Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Workspace/Solution/Checksum.cs:143]: I hoenstly don't understand the design between checksum/hashdata.  the former seems to exist just to add a heap allocation on every checksum we have.

Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Workspace/Solution/Checksum.cs:143]: no clue.  it's a bit of a wonky design imo.

Comment (sharwell) [reply src/Workspaces/Core/Portable/Workspace/Solution/Checksum.cs:143]: I would prefer an approach that doesn't further expose `HashData` (ideally we make the type `private`)

Comment (sharwell) [reply src/Workspaces/Core/Portable/Workspace/Solution/Checksum.cs:143]: Huh... just realized `Checksum` is a reference type. How unfortunate.

**Review by sharwell — APPROVED** [review]
Approved for 16.10

**Pull request merged**